nntrainer/include/network_graph.h

```diff
@@ -0,0 +1,81 @@
+// SPDX-License-Identifier: Apache-2.0
+/**
+ * Copyright (C) 2020 Jijoong Moon <jijoong.moon@samsung.com>
+ *
+ * @file    network_graph.h
+ * @date    19 Oct 2020
+ * @see     https://github.com/nnstreamer/nntrainer
+ * @author  Jijoong Moon <jijoong.moon@samsung.com>
+ * @bug     No known bugs except for NYI items
+ * @brief   This is Network Graph Class for Neural Network
+ *
+ */
+
+#ifndef __NETWORK_GRAPH_H_
+#define __NETWORK_GRAPH_H_
+#ifdef __cplusplus
+
+#include <iostream>
+#include <layer_internal.h>
+#include <list>
+#include <loss_layer.h>
+#include <memory>
+#include <stack>
+#include <vector>
+
+namespace nntrainer {
+
+struct LayerNode {
+  std::shared_ptr<Layer> layer;
+  unsigned int index;
+};
+
+class NetworkGraph {
+public:
+  NetworkGraph() : num_node(0), def_name_count(0){};
+
+  void addEdge(unsigned int ith, LayerNode node);
+
+  void topologicalSort();
+
+  void addLayerNode(std::shared_ptr<Layer>);
+
+  unsigned int getNumNode() { return num_node; }
+
+  LayerNode &getLayerNode(unsigned int ith);
+
+  LayerNode &getLayerNode(const std::string layer_name);
+
+  void ensureName(std::shared_ptr<Layer> layer, const std::string &prefix,
+                  bool force_rename);
+
+  int setGraphNode(std::vector<std::shared_ptr<Layer>> layers,
+                   const LossType loss_type);
+
+  int realizeMultiInputType(Layer &current);
+
+  int realizeActivationType(Layer &current,
```

Comment:
Also this is a suggestion to modularize this - separate this into two classes -  let says `GraphCore` which only does graph related stuff like `addEdge`, `addNode`, `topologicalSort` and make this class templated.
Then `NetworkGraph` class will have an instance of `GraphCore<LayerNode>` and will directly use it, making it easier to manage graph and these realization functions.
This will also make commits smaller and easier to review.

Reply:
I will do it later. Cause the model create layer in model_loader, and there is layers vector. at this time, each layers are not initialized. In stead of create layers as an vector, we could create GraphCore directly.

Reply:
Yes, later works too but this decoupling is needed.

Reply:
> Yes, later works too but this decoupling is needed.

Agreed. Which is oop does.